### PR TITLE
docs: Documentation on Flask-SQLAlchemy

### LINKED ISSUE #786: A blog talking and/or teaching about Flask-SQAlchemy and how the Mentorship Backend uses it with example user models.
**Description**
As a contributor, I need you to come up with a blog or a file in .md teaching about Flask-SQAlchemy and how it is used in this project. An explanatory report will be appreciated not only teaching from the basics but also describing it's full usage here.

**Acceptance Criteria**

- [ ] Update required

- [ ] Submit as a markdown file

**Definition of Done**

- [ ] All of the required items are completed.

- [ ] Approval by 2 mentors.

**Estimation**
4 to 5 hours

**Note**
This issue is can be contributed to only from 1st September when Open Source Hack starts.

### PR DESCRIPTION
**Pull request by ashwinijha6**
### Description
Documentation talking & teaching about Flask-SQAlchemy and how the Mentorship Backend uses it

Fixes #786

### Type of Change:

- Code
- Quality Assurance
- Outreach
- Documentation

**Code/Quality Assurance Only**
- Bug fix (non-breaking change which fixes an issue)
- This change requires a documentation update (software upgrade on readme file)
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] Any dependent changes have been merged
- [ ] Update Postman API at /docs folder
- [ ] Update Swagger documentation and the exported file at /docs folder
- [ ] Update requirements.txt

**Code/Quality Assurance Only**

- [x] My changes generate no new warnings 
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been published in downstream modules

### CODE REVIEW

**Comment by ashwinijha6** [comment]
I have made the required changes in this pull request @SanketDG

**Review by yashasvimisra2798 — DISMISSED** [review]
LGTM

**Comment by ashwinijha6** [comment]
@keshakaneria I need another review on this pull request I have made the changes suggested by @SanketDG . Could u pls review my pull request.

**Review by keshakaneria — DISMISSED** [review]
Good work!

**Comment by ashwinijha6** [comment]
> Good work!

Thanku @keshakaneria ❤️

**Comment by SanketDG** [comment]
Pinging @isabelcosta for a second set of eyes on this, considering this is very extensive.

**Comment by rpattath** [comment]
Changing this PR to `Needs Review` because @isabelcosta wants to give it a read. Once she approves, this PR can go to `Ready to Merge` it does not need testing.

**Comment by ashwinijha6** [comment]
> Changing this PR to `Needs Review` because @isabelcosta wants to give it a read. Once she approves, this PR can go to `Ready to Merge` it does not need testing.

Ok @rpattath  sure

File: docs/FlaskSQLAlchemy.md
Comment (devkapilbansal) [inline docs/FlaskSQLAlchemy.md]: Why so many blank lines here??

Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md]: I didn't notice it in the doc should I resolve it

Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md]: ?

Comment (devkapilbansal) [reply docs/FlaskSQLAlchemy.md]: Yes, you should

Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md]: Ok I will

**Comment by ashwinijha6** [comment]
@devkapilbansal l removed the extra lines & also squashed the commits pls review

**Comment by ashwinijha6** [comment]
@keshakaneria & @yashasvimisra2798  pls review my pull request I have made the changes

File: docs/FlaskSQLAlchemy.md
Comment (devkapilbansal) [inline docs/FlaskSQLAlchemy.md]: Probably what you want is not happening here.  GitHub is not formatting this document as it should. 
Add a blank line here if you want it in new line
error

**Review by devkapilbansal — CHANGES_REQUESTED** [review]

File: docs/FlaskSQLAlchemy.md
Comment (devkapilbansal) [reply docs/FlaskSQLAlchemy.md]: Image taken from :- https://github.com/ashwinijha6/mentorship-backend/blob/flask/docs/FlaskSQLAlchemy.md

Comment (keshakaneria) [reply docs/FlaskSQLAlchemy.md]: @ashwinijha6 you can better use [` `] code block rather than [**] to bold your text.
For example:
For a relative file path, this requires three slashes `sqlite://<nohostname>/<path>` where `<path>` is relative

Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md]: ok I will try that

**Comment by ashwinijha6** [comment]
**sqlite://`<`nohostname`>`/`<`path`>`**
@devkapilbansal & @keshakaneria could u'll check with this, it is giving me the correct output but highlighting <>
Can it be used?

**Comment by ashwinijha6** [comment]
Also I guess the two lines related to the relative path & absolute path be just aligned beneath the SQLite line what do u'll think?

File: docs/FlaskSQLAlchemy.md
Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md]: How did u use code blocks because I'm getting it wrong within code blocks as well

Comment (keshakaneria) [reply docs/FlaskSQLAlchemy.md]: Write the code inside this ``

**Comment by keshakaneria** [comment]
> **sqlite://`<`nohostname`>`/`<`path`>`**
> @devkapilbansal & @keshakaneria could u'll check with this, it is giving me the correct output but highlighting <>
> Can it be used?

You have ended the code block symbol after the < itself. Start the code block when you want to start and end only when the code block is completed. ex.  "sqlite://(nohostname)/(path)" (replace `""` with ` ` and `(` with ` < ` because if i will write the github editor will change it into code block 😛)
Also the editor skips  `<` portion so we can't put the statement inside anything other than code blocks using `` just before when we want to start and end ONLY when we want the code block to end.

**Comment by ashwinijha6** [comment]
> > **sqlite://`<`nohostname`>`/`<`path`>`**
> > @devkapilbansal & @keshakaneria could u'll check with this, it is giving me the correct output but highlighting <>
> > Can it be used?
> 
> You have ended the code block symbol after the < itself. Start the code block when you want to start and end only when the code block is completed. ex. "sqlite://(nohostname)/(path)" (replace `""` with ` ` and `(` with `<` because if i will write the github editor will change it into code block 😛)
> Also the editor skips `<` portion so we can't put the statement inside anything other than code blocks using `` just before when we want to start and end ONLY when we want the code block to end.

thanx @keshakaneria I got it

**Comment by ashwinijha6** [comment]
@keshakaneria & @devkapilbansal could u'll just check once then I will squash the changes

File: docs/FlaskSQLAlchemy.md
Comment (devkapilbansal) [inline docs/FlaskSQLAlchemy.md]: Why you wrote Python 3.7 in SQLAlchemy column?

File: docs/FlaskSQLAlchemy.md
Comment (devkapilbansal) [inline docs/FlaskSQLAlchemy.md:42]: Do we need to give instructions for installing Python too?

File: docs/FlaskSQLAlchemy.md
Comment (devkapilbansal) [inline docs/FlaskSQLAlchemy.md:102]: It should be ``` load the configuration of your choice ``` or something like that

**Review by devkapilbansal — CHANGES_REQUESTED** [review]

File: docs/FlaskSQLAlchemy.md
Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md]: https://flask-sqlalchemy.palletsprojects.com/en/2.x/
have a look here once I don't exactly know what this means

File: docs/FlaskSQLAlchemy.md
Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md:42]: I have written the doc from the perspective of a new comer that why I have mentioned it

File: docs/FlaskSQLAlchemy.md
Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md:102]: I guess **of choice** is also correct 
of choice means from the list that the link denotes

File: docs/FlaskSQLAlchemy.md
Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md]: I guess it means that SQLAlchemy of version 1.0.10+ is compatible with Python 3.7

**Comment by ashwinijha6** [comment]
@keshakaneria could you pls review my pull request

File: docs/FlaskSQLAlchemy.md
Comment (keshakaneria) [inline docs/FlaskSQLAlchemy.md:110]: There are 4 slashes still. Did you miss something or is it this way?

File: docs/FlaskSQLAlchemy.md
Comment (keshakaneria) [inline docs/FlaskSQLAlchemy.md:143]: Here too

File: docs/FlaskSQLAlchemy.md
Comment (keshakaneria) [inline docs/FlaskSQLAlchemy.md:151]: And here also

File: docs/FlaskSQLAlchemy.md
Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md:110]: Yes it is that way only for MAC/UNIX 4 slashes are required

File: docs/FlaskSQLAlchemy.md
Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md:151]: I have mentioned the 4 slashes part here

File: docs/FlaskSQLAlchemy.md
Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md:143]: As I used the above configuration that's why I explained it here as well

File: docs/FlaskSQLAlchemy.md
Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md:42]: What do u think should I change this?

Comment (devkapilbansal) [reply docs/FlaskSQLAlchemy.md:42]: Actually I don't know exactly. We should wait for some more reviewers to review it and comment.

File: docs/FlaskSQLAlchemy.md
Comment (devkapilbansal) [reply docs/FlaskSQLAlchemy.md:102]: Okay, cool then

File: docs/FlaskSQLAlchemy.md
Comment (keshakaneria) [reply docs/FlaskSQLAlchemy.md:151]: Alright then its fine!

**Comment by ashwinijha6** [comment]
@keshakaneria I feel that the python installation part should be there as it will be easy for the newbie's for creating a flask application what do u think?

File: docs/FlaskSQLAlchemy.md
Comment (devkapilbansal) [reply docs/FlaskSQLAlchemy.md]: But Python versions are written in another column which are compatible with it. You should look into it once again

**Comment by keshakaneria** [comment]
> @keshakaneria I feel that the python installation part should be there as it will be easy for the newbie's for creating a flask application what do u think?

Yeah that part is well i suppose 👍

**Comment by ashwinijha6** [comment]
> > @keshakaneria I feel that the python installation part should be there as it will be easy for the newbie's for creating a flask application what do u think?
> 
> Yeah that part is well i suppose 👍

ok 👍

File: docs/FlaskSQLAlchemy.md
Comment (keshakaneria) [reply docs/FlaskSQLAlchemy.md:42]: Yeah this is fine i suppose, it will be beneficial for new comers

File: docs/FlaskSQLAlchemy.md
Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md:102]: Should I change this?

File: docs/FlaskSQLAlchemy.md
Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md]: Did u see the link?

Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md]: Ok I will look into it again

Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md]: I think it is just in the SQLAlchemy column because Python 3.7 is compatible with SQLAlchemy 1.0.10+ 
I will remove it. What do u think?

**Comment by ashwinijha6** [comment]
@devkapilbansal & @keshakaneria The Requirements table containing SQLAlchemy 1.0.10+ w/ Python 3.7, I guess it only means SQLAlchemy version 1.0.10 with Python 3.7 & that's for that specific version so if u'll suggest I will remove it from the table?

This Requirements table is in the documentation

**Comment by ashwinijha6** [comment]
@devkapilbansal I have made the changes could u pls review it

File: docs/FlaskSQLAlchemy.md
Comment (devkapilbansal) [inline docs/FlaskSQLAlchemy.md:352]: @ashwinijha6 @keshakaneria instead of writng it like this, wouldn't it be better if we paste link of those lines here
LIke: `https://github.com/anitab-org/mentorship-backend/blob/eae562267b1cbbb597ea3a8829b87ad373fa323c/config.py#L98-103`
https://github.com/anitab-org/mentorship-backend/blob/eae562267b1cbbb597ea3a8829b87ad373fa323c/config.py#L98-L103

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
Great work @ashwinijha6 squash your commit and see my final request

File: docs/FlaskSQLAlchemy.md
Comment (devkapilbansal) [reply docs/FlaskSQLAlchemy.md:352]: Also, do we need to write flask env config and then sqlalchemy config at different places. May be both are same. Although I am not sure about this

**Review by keshakaneria — DISMISSED** [review]
Looks good to me right now 😄! Just squash the commits after final request.

File: docs/FlaskSQLAlchemy.md
Comment (keshakaneria) [reply docs/FlaskSQLAlchemy.md:352]: That is good over there i suppose

Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md:352]: @devkapilbansal I guess I have mentioned the link from where the code is taken
Do u want me to mention from line 98 to **103** means the 103?
Also I have mentioned the code for better understanding because the issue stated where sqlalchemy has been used in mentorship backend

Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md:352]: @devkapilbansal could u pls state the exact line where I have used flask env config & sqlalchemy config both

Comment (devkapilbansal) [reply docs/FlaskSQLAlchemy.md:352]: I mean that instead of writing the config part by yourself give its link.

Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md:352]: @devkapilbansal  I have given the link

**Comment by ashwinijha6** [comment]
link

File: docs/FlaskSQLAlchemy.md
Comment (devkapilbansal) [reply docs/FlaskSQLAlchemy.md:352]: Leave it. I tried it manually but seems it work in comments only, not in markdown file :confused:

**Review by devkapilbansal — DISMISSED** [review]
Thanks @ashwinijha6 just squash your commits in one
LGTM :+1:

**Comment by devkapilbansal** [comment]
Pinging @isabelcosta to give it a read

**Comment by vj-codes** [comment]
Adding label ready to merge as it's a documentation PR

**Comment by devkapilbansal** [comment]
@vj-codes ready to merge was removed because @isabelcosta wants to read it once.
See the comment above:-
https://github.com/anitab-org/mentorship-backend/pull/894#issuecomment-707910486

File: docs/FlaskSQLAlchemy.md
Comment (ashwinijha6) [reply docs/FlaskSQLAlchemy.md:352]: yes https://github.com/anitab-org/mentorship-backend/blob/eae562267b1cbbb597ea3a8829b87ad373fa323c/config.py#L98-103
just reflects to the line 98 not the entire code block 98-103

**Comment by ashwinijha6** [comment]
@devkapilbansal & @keshakaneria I have squashed the commits

**Review by devkapilbansal — APPROVED** [review]
Thanks, @ashwinijha6 .LGTM :+1:

**Review by keshakaneria — APPROVED** [review]
Perfect 😄

**Comment by ashwinijha6** [comment]
@isabelcosta could u pls review this doc

**Review by vj-codes — APPROVED** [review]
LGTM 🎉 
Thank you for your contribution @ashwinijha6

**Comment by isabelcosta** [comment]
@ashwinijha6 thank you for writing this extensive document 🙌🏾 
In our open session, maintainers decided to close this issue, as this content, would better fit a blog post perhaps since it's more related to the technology itself then the app.

**Pull request closed without merging**
